Bugsnag/Helpers/BSGInternalErrorReporter.m

```diff
@@ -0,0 +1,177 @@
+//
+//  BSGInternalErrorReporter.m
+//  Bugsnag
+//
+//  Created by Nick Dowell on 06/05/2021.
+//  Copyright © 2021 Bugsnag Inc. All rights reserved.
+//
+
+#import "BSGInternalErrorReporter.h"
+
+#import "BSG_KSSystemInfo.h"
+#import "BSG_RFC3339DateTool.h"
+#import "BugsnagApiClient.h"
+#import "BugsnagAppWithState+Private.h"
+#import "BugsnagCollections.h"
+#import "BugsnagConfiguration+Private.h"
+#import "BugsnagDeviceWithState+Private.h"
+#import "BugsnagError+Private.h"
+#import "BugsnagEvent+Private.h"
+#import "BugsnagHandledState.h"
+#import "BugsnagKeys.h"
+#import "BugsnagLogger.h"
+#import "BugsnagMetadata+Private.h"
+#import "BugsnagNotifier.h"
+#import "BugsnagStackframe+Private.h"
+#import "BugsnagUser+Private.h"
+
+static NSString * const EventPayloadVersion = @"4.0";
+
+static NSString * const BugsnagDiagnosticsKey = @"BugsnagDiagnostics";
+
+BugsnagHTTPHeaderName const BugsnagHTTPHeaderNameInternalError = @"Bugsnag-Internal-Error";
+
+
+NSString *BSGErrorDescription(NSError *error) {
+    return [NSString stringWithFormat:@"%@ %ld: %@", error.domain, (long)error.code,
+            error.userInfo[NSDebugDescriptionErrorKey] ?: error.localizedDescription];
+}
+
+
+// MARK: -
+
+@interface BSGInternalErrorReporter ()
+
+@property (weak, nullable, nonatomic) id<BSGInternalErrorReporterDataSource> dataSource;
+@property (nonatomic) NSURLSession *session;
+
+@end
+
+
+@implementation BSGInternalErrorReporter
+
+static BSGInternalErrorReporter *sharedInstance_;
+
++ (BSGInternalErrorReporter *)sharedInstance {
+    return sharedInstance_;
+}
+
++ (void)setSharedInstance:(BSGInternalErrorReporter *)sharedInstance {
+    sharedInstance_ = sharedInstance;
+}
+
+- (instancetype)initWithDataSource:(id<BSGInternalErrorReporterDataSource>)dataSource {
+    if ((self = [super init])) {
+        _dataSource = dataSource;
+        _session = [NSURLSession sessionWithConfiguration:NSURLSessionConfiguration.ephemeralSessionConfiguration];
+    }
+    return self;
+}
+
+// MARK: Public API
+
+- (void)reportErrorWithClass:(NSString *)errorClass
+                     message:(nullable NSString *)message
+                 diagnostics:(nullable NSDictionary<NSString *, id> *)diagnostics {
+    @try {
+        BugsnagEvent *event = [self eventWithErrorClass:errorClass message:message diagnostics:diagnostics];
+        if (event) {
+            [self sendEvent:event];
+        }
+    } @catch (NSException *exception) {
+        bsg_log_err(@"%@", exception);
+    }
+}
+
+// MARK: Private API
+
+- (nullable BugsnagEvent *)eventWithErrorClass:(NSString *)errorClass
+                                       message:(nullable NSString *)message
+                                   diagnostics:(nullable NSDictionary<NSString *, id> *)diagnostics {
+    id<BSGInternalErrorReporterDataSource> dataSource = self.dataSource;
+    if (!dataSource) {
```

Comment:
Should we even allow dataSource to be nil?

Reply:
It's a weak reference so cannot be guaranteed to be nonnull.